File: public/exports/spec/06_scripts/signature-demo/verify.ts
import fs from 'fs'
import crypto from 'crypto'

const [feedPath, publicKeyPath] = process.argv.slice(2)
if (!feedPath || !publicKeyPath) {
  console.error('Usage: tsx verify.ts <feed.json> <public.pem>')
  process.exit(1)
}

const feed = JSON.parse(fs.readFileSync(feedPath, 'utf-8'))
const publicKeyPem = fs.readFileSync(publicKeyPath, 'utf-8')
const publicKey = crypto.createPublicKey(publicKeyPem)

const sig = feed.signature
if (!sig || sig.algorithm !== "ed25519") {
  throw new Error("Missing or unsupported signature format (expecting { algorithm: 'ed25519', value })")
}

const signature = Buffer.from(sig.value, 'base64')
const blocks = feed.trust.signed_blocks
const dataToVerify: Record<string, any> = {}
for (const key of blocks) {
  dataToVerify[key] = feed[key]
}

const payload = JSON.stringify(dataToVerify, Object.keys(dataToVerify).sort(), 0)
const isValid = crypto.verify(null, Buffer.from(payload), publicKey, signature)

console.log("✅ Signature valid?", isValid)
if (!isValid) process.exit(1)

File: public/exports/spec/06_scripts/signature-demo/verify.py
import json
import sys
import base64
from pathlib import Path
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

if len(sys.argv) != 3:
    print("Usage: python verify.py <feed.json> <public.pem>")
    sys.exit(1)

feed_path = Path(sys.argv[1])
pubkey_path = Path(sys.argv[2])

with feed_path.open("r", encoding="utf-8") as f:
    feed = json.load(f)

with pubkey_path.open("rb") as f:
    public_key = serialization.load_pem_public_key(f.read())

sig = feed.get("signature")
if not sig or not isinstance(sig, dict) or sig.get("algorithm") != "ed25519":
    raise ValueError("Missing or invalid 'signature' block (expected { algorithm: 'ed25519', value })")

signature = base64.b64decode(sig["value"])

blocks = feed["trust"]["signed_blocks"]
data_to_verify = {k: feed[k] for k in blocks}
payload = json.dumps(data_to_verify, sort_keys=True, separators=(",", ":")).encode("utf-8")

try:
    public_key.verify(signature, payload)
    print("✅ Signature is valid.")
except Exception as e:
    print("❌ Signature is INVALID.")
    print(str(e))
    sys.exit(1)
